File: Python/hoar_sort.py
# average is O(nlogn), the worst time is O(n ** 2)
def hoar_sort(array):
    if len(array) <= 1:    # first of all we need to define extreme case, whick is when array length less than two
        return
    barrier = array[0]  # second step is to find a barrier element, since we don't know either this array sorted or not, we will take first element
    L, M, R = [], [], []   # also we need three list where we can append elements whick are less, equal or greater than barries element
    for x in array:   # in this cycle we can distribute elements into lists
    
        if x < barrier:
            L.append(x)
        elif x == barrier:
            M.append(x)
        else:
            R.append(x)
    hoar_sort(L)   # sort of lower elements
    hoar_sort(R)    # sort of greater elements
    count = 0   # since we don't want to return new list, we need to change old list, so that we need index
    for el in L + M + R:
        array[count] = el
        count += 1



def check_sorted(array, ascending=True):    # also we need to check if array is sorted
    flag = True     # first of all we need flag which shows either array sorted or not
    s = 2 * int(ascending) - 1  # second is indicator which helps us not to duplicate code, it will be 1 is ascending true otherwise -1
    for i in range(0, len(array)-1):    
        if s * array[i] > s * array[i+1]:   #checking if curret element is greater than next, so that we can return false right now
            flag = False
            return flag
    return flag 






if (__name__ == "__main__"):
    list1 = [2, 4, 5, 6, 78, 2, 1, 8]
    list2 = [1]
    list3 = []
    hoar_sort(list1)
    hoar_sort(list2)
    hoar_sort(list3)
    
    print(check_sorted(list1))
    print(check_sorted(list2))
    print(check_sorted(list3))

File: Java/HoarSort.java
import java.util.Arrays;

public class HoarSort {
    public static void main(String[] args) {

        int[] ar1 = {4, 5, 6, 1, 4, 2, 1, 4, 6, 7, 8};
        hoarSort(ar1, 0, 10);
        Arrays.stream(ar1).forEach(el->System.out.print(el + " "));
        System.out.println();
        int[] ar2 = {5, 5, 5, 5, 5};
        hoarSort(ar2, 0, 4);
        Arrays.stream(ar2).forEach(el->System.out.print(el + " "));
        System.out.println();
        int[] ar3 = {-1, -2, -3, 7};
        hoarSort(ar3, 0, 3);
        Arrays.stream(ar3).forEach(el->System.out.print(el + " "));
        System.out.println();
        int[] ar4 = {};
        hoarSort(ar4, 0, 0);
        Arrays.stream(ar4).forEach(el->System.out.print(el + " "));
        System.out.println();
        int[] ar5 = {1, 2, 3, 4, 5};
        hoarSort(ar5, 0, 4);
        Arrays.stream(ar5).forEach(el->System.out.print(el + " "));
        System.out.println();
        int[] ar6 = {5,3,1,2,2,1,0};
        hoarSort(ar6, 0, 6);
        Arrays.stream(ar6).forEach(el->System.out.print(el + " "));


    }

    public static void hoarSort(int[] array, int start, int end) {
        // start is the first element's index and end is the last element's index
        if (start >= end) return;  //checking the end of recursion
        int breakPoint = hoarPartition(array, start, end);   // here we can divide our array
        // we don't need to replace element which has breakPoint index since it's already on it's place
        hoarSort(array, start, breakPoint - 1);   // here we sort left part of array
        hoarSort(array, breakPoint + 1, end);   //here we sort right part of array
    }


    public static int hoarPartition(int[] array, int start, int end) {
        int i = start + 1;  //we're gonna start our partition from the element with this index
        int barrier = array[start]; //this is the index of out barrier element
        int j = end;
        while (true) {
            while (i < end && array[i] < barrier) i++;   // searching for the element in the left side which is greater than barrier or equals to it 
            while (array[j] > barrier) j--;   //searching for the element which is less than barrier in the right side 
            if(i >= j) break;
            
            swap(array, i++, j--);  //swapping element since i element is greater than j element
        
        };
        swap(array, start, j);    // putting barrier element to it's place
        return j;
    }



    public static void swap(int[] array, int i, int j) {
        // swapping two elements
        int temp = array[i];
        array[i] = array[j];
        array[j] = temp;
    }

    }
        

 
